Refactor CI avoiding test circular dependencies

vdukhovni opened the pull request:
Instead of running tests in the parent "bytestring" project, with
circular dependencies from the test frameworks, build and run them in
the child `bytestring-tests` project, which builds the parent as a
dependency that does not run tests, avoiding all the loops.

This requires cabal-install >= 2.4 with cabal.project doing most of the
heavy lifting.

vdukhovni commented:
@sjakobi @hvr @cartazio @bgamari @Bodigrim
I believe this is ready and makes testing substantially easier going forward.

cartazio commented:
Looks great to me!

On Wed, May 27, 2020 at 3:26 PM Viktor Dukhovni <notifications@github.com>
wrote:

> @sjakobi <https://github.com/sjakobi> @hvr <https://github.com/hvr>
> @cartazio <https://github.com/cartazio> @bgamari
> <https://github.com/bgamari> @Bodigrim <https://github.com/Bodigrim>
> I believe this is ready and makes testing substantially easier going
> forward.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/bytestring/pull/227#issuecomment-634890514>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABBQQDOSW5LXA4ULAHWETRTVSQDANCNFSM4NMNYHUQ>
> .
>

vdukhovni commented:
> Looks great to me!

Thanks.  The one thing that's odd is I'm not seeing a Travis job for this PR.  Don't know why...

sjakobi commented on tests/bytestring-tests.cabal:
```diff
@@ -44,6 +54,10 @@ executable prop-compiled
 
 executable regressions
```
Shouldn't this also be a testsuite?

vdukhovni replied:
It hasn't been testable in ages.  It tries to allocate more RAM than fits in an `Int`, but that's no longer possible.

sjakobi commented on tests/cabal.project:
```diff
@@ -0,0 +1,9 @@
+           , *:ghc-prim
+           , regex-posix:base
+
+package bytesting-tests
+    tests: True
```
Why not

```suggestion
tests: True
```

?

vdukhovni replied:
We don't want to test any of our dependencies.  I am not sure what the scope of that is, it may cover dependencies.

sjakobi replied:
The default scope are the packages listed above AFAIK.

sjakobi commented on cabal.project:
```diff
@@ -0,0 +1,5 @@
+packages: bytestring.cabal
+allow-newer: *:bytestring
+
+tests: False
+benchmarks: False
```
Why is this necessary when there aren't any testsuites or benchmarks?

vdukhovni replied:
Probably redundant, but harmless. Belt and suspers, but the belt is quite strong...

sjakobi commented on cabal.project:
```diff
@@ -0,0 +1,5 @@
+packages: bytestring.cabal
+allow-newer: *:bytestring
```
Surely there can't be any reverse dependencies of bytestring involved here?

vdukhovni replied:
When testing with GHC-head, `text`, et. al. needed for the tests want an older `bytestring` than the one we're testing, but we want to force them to use the new one.

sjakobi replied:
But this `cabal.project` doesn't include the tests?!

sjakobi commented on .travis.yml:
```diff
@@ -2,20 +2,22 @@ env:
  # https://github.com/haskell/bytestring/issues/213
+ # With Cabal 2.4 we also get improved syntax in cabal.project files,
+ # supporting remote sources, and other goodies, so use it across the
+ # board.  This branch is not GHC 7.x compatible, so 8.x-only.
```
```suggestion
 # board.
```

vdukhovni replied:
Thanks, will do.

sjakobi replied:
Thanks! This is resolved then, but I don't have the button to mark it… :/

sjakobi commented on .travis.yml:
```diff
@@ -2,20 +2,22 @@ env:
  # Note: We have to use Cabal 2.4 with GHC 7.0 since for some reason
  # builds of "text" fail with GHC 7.0 and Cabal versions prior to 2.4
  # https://github.com/haskell/bytestring/issues/213
```
Nit: Maybe remove this and add a reference to the issue in the link below

vdukhovni replied:
Not sure what you want here, can suggest a diff?

sjakobi replied:
Apologies, please mark resolved. Not sure where I was going there anymore. :/

sjakobi commented on .travis.yml:
```diff
@@ -32,31 +34,20 @@ before_install:
+ - cabal v2-sdist -o .
+ # Remove to enable build of sdist.
+ - rm -f cabal.project
+ - rm -f tests/test-builder.tix
+ - (cd tests; cabal v2-configure; cabal v2-test)
  # "cabal check" disabled due to -O2 warning
- - export SRC_TGZ=$(cabal info . | awk '{print $2 ".tar.gz";exit}') ;
-   cd dist/;
-   if [ -f "$SRC_TGZ" ]; then
-      cabal install --force-reinstalls "$SRC_TGZ";
+ - export SRC=$(cabal info . | awk '{print $2; exit}');
+   if [ -f "$SRC.tar.gz" ]; then
+      cabal get "./$SRC.tar.gz";
+      (cd "$SRC"; cabal install --force-reinstalls .);
```
Did you consider simply using https://github.com/haskell-CI/haskell-ci to generate this script?

sjakobi replied:
Actually, that might be tricky…

I can give this a spin once this PR has been merged.

vdukhovni replied:
I'm having a devil of a time getting it to work portably between Cabal 2.4, 3.0 and 3.2, not sure the autogenerated one would fare better.

vdukhovni replied:
Got it to work, but had to resort to `v1-install` with Cabal 3.x, if you figure out a working, non-deprecated syntax, I'm all ears...

sjakobi replied:
Compatibility issues like that are a `haskell-ci` specialty. ;)

No worries though. Looks like CI is finally green now!

sjakobi commented on tests/cabal.project:
```diff
@@ -0,0 +1,9 @@
+packages: bytestring-tests.cabal
+          ../bytestring.cabal
```
Why include the `bytestring` package here? Doesn't this cause test dependencies to be recompiled after changes to bytestring code?

vdukhovni replied:
IIRC the test frameworks and text need to use the same version of ByteString as being tested. We compiling the tests with code that also uses Text, which uses ByteString, ... that cyclic thing.  But here we're build the test module, and `bytestring` is just one of the dependencies to build along with the others.

Give it a try:
```
cd tests
cabal configure
cabal test
```
then tweak `cabal.project` and try again...

sjakobi commented:
Many thanks for responding to my comments above, @vdukhovni! I have, in turn, tried to respond to your responses, but I don't have the necessary to permissions to "unresolve". Would you mind unresolving these things for me?

Alternatively, maybe @hvr or @cartazio could give me the necessary permissions. That would be greatly appreciated! :)

vdukhovni commented:
I've `unresolved` all the previous comments, which I marked "resolved" primarily so I could see what I had not yet addressed, otherwise hard to find... Sorry...

sjakobi commented:
I just gave this PR as spin on my local machine. The issue that I believe we're close to eliminating now is that after any change to the `bytestring` library code, the entire set of test dependencies is rebuilt, which takes a bit of time. A development ergonomics problem, IMHO.

What we need to do is to keep the local `bytestring` _package_ out of the `bytestring-test`.

At this point, that's very simple to achieve:

1. Revert 84f15d5 (I really don't understand where you wanted to go with that @vdukhovni)
2. Remove the local `bytestring` package from the test `cabal.project`:

```patch
--- a/tests/cabal.project
+++ b/tests/cabal.project
@@ -1,5 +1,4 @@
 packages: bytestring-tests.cabal
-          ../bytestring.cabal

 allow-newer: *:bytestring
            , *:ghc-prim
```

Now, rebuilding the tests is super-fast. Great!

sjakobi commented:
(And thanks for unresolving my earlier comments! It's a bit of a shame that we have to jump through these hoops to collaborate!)

vdukhovni commented:
> I just gave this PR as spin on my local machine. The issue that I believe we're close to eliminating now is that after any change to the `bytestring` library code, the entire set of test dependencies is rebuilt, which takes a bit of time. A development ergonomics problem, IMHO.

Rebuilding the test dependencies against the updated `bytestring` is actually *necessary*. Otherwise the test dependencies are linked against the wrong `bytestring`.  The time to run the tests while making changes to `bytestring` should not influence the decision much or at all.  It is pretty quick anyway.

> What we need to do is to keep the local `bytestring` _package_ out of the `bytestring-test`.

No, we actually need to make sure that we're testing the local `bytestring` package and not the stock `bytestring` from the package cache.

> At this point, that's very simple to achieve:
> 
> 1. Revert [84f15d5](https://github.com/haskell/bytestring/commit/84f15d5d21001c3a2de458df586e3a5ebea637e2) (I really don't understand where you wanted to go with that @vdukhovni)
> 2. Remove the local `bytestring` package from the test `cabal.project`:

I disagree.  Please try for example building Andrew Martin's PR, or the `ryu` PR without this.  (For the former you'll need GHC master, and Simon PJ's patch), ...

> Now, rebuilding the tests is super-fast. Great!

I disagree, and don't find the time to build tests a deciding factor.

vdukhovni commented:
> Note that the `containers` setup takes this idea even further. It's `containers-tests` sub-package simply has the same library component as `containers` proper. That way, modules used by multiple testsuites are recompiled only once.

Are you reachable via any interactive channels?  (Skype, Hangouts, Signal)?  Please send me an email <ietf-dane@dukhovni.org>.  For me the test runtime from cold start was:
```
real    1m57.371s
user    2m44.350s
sys     0m16.233s
```
Which is not a barrier to running tests now and then...  Builds w/o tests take 16s (cold).

sjakobi commented:
I've sent you an email.

My concern is about interactive development. You're working on some function, and you want to run the tests for that. Recompilation should be as fast as possible then. 1 or  even 2 minutes would be very frustrating already.

vdukhovni commented:
With great help from @sjakobi I've simplified this further, and expect I've addressed his concerns. The CI should shortly show up green. The cabal.project files are now quite short, and we're no longer rebuilding the test dependencies against the new `bytestring`, just testing `bytestring-test` as suggested.

The only downside is that modules need to be listed in two .cabal files, `bytestring` and `bytestring-test`

sjakobi commented:
Thanks @vdukhovni. I very much enjoyed our chat! :)

FYI, I believe I've found the discussion on the `cabal` feature that would make much of this work redundant by allowing circular dependencies in testsuites: https://github.com/haskell/cabal/issues/4087

vdukhovni commented:
> FYI, I believe I've found the discussion on the `cabal` feature that would make much of this work redundant by allowing circular dependencies in testsuites: [haskell/cabal#4087](https://github.com/haskell/cabal/issues/4087)

Yeah, that would have been nice.  But it's been under discussion for 3 years, so I'm not expecting a miracle...  I've had to just force push a small fix, it seems I needed a `v2-build` rather than `build` for testing that that the unpacked `sdist` is buildable...

sjakobi commented on .travis.yml:
```diff
@@ -32,31 +35,20 @@ before_install:
+ - cabal v2-sdist -o .
+ # Remove to enable build of sdist.
+ - rm -f cabal.project
+ - rm -f tests/test-builder.tix
```
This should be redundant after the removal of the `-fhpc` flag?

```suggestion
```

vdukhovni replied:
Yes, but also harmless, if the -fhpc comes back, but if you'd like to tidy it up, I can remove it...

sjakobi replied:
I'd slightly prefer to have this removed.

vdukhovni commented:
> It would be great if someone with better bash skills could review the Travis script.

I guess, sadly, my own `bash` power-user status does not count, you're looking for someone *else* to review...

vdukhovni commented:
@hvr, @cartazio.  At this point I'm looking for a final review and merge, after addressing any issues.

sjakobi commented:
> > It would be great if someone with better bash skills could review the Travis script.
> 
> I guess, sadly, my own `bash` power-user status does not count, you're looking for someone _else_ to review...

Indeed. I'm looking for a second pair of eyeballs in addition to yours here. My own count for at most half an eye… ,)

bgamari commented:
@hvr, @cartazio, @dcoutts, this is in need of review.

vdukhovni commented:
Rebased as mentioned in #214

sjakobi commented:
@Bodigrim, @hsyl20, could you take a look at this?

I'm not too worried about potential problems with this. If any issue with CI comes up, we can fix it on the go.

If no one points out any problems with this patch, I'd simply merge it in 2 or 3 days.

hsyl20 commented:
Looks good to me too.

sjakobi commented:
Thank you, @vdukhovni! :)

vdukhovni commented on .travis.yml:
```diff
@@ -32,31 +35,19 @@ before_install:
+      cabal get "./$SRC.tar.gz";
+      (cd "$SRC"; cabal v2-build);
    else
       echo "expected '$SRC_TGZ' not found";
```
Oops, the error message here is not quite right, it should be `$SRC.tar.gz` not `$SRC_TGZ` which is no longer set.  This should not happen, but if it ever did, the error message would be confusing...

vdukhovni replied:
@sjakobi Would you care to fix the error message?

sjakobi replied:
I'm not too worried about it. Feel free to send a PR or open an issue.
